File: 0721-accounts-merge/0721-accounts-merge.java
class Solution {
    private Map<String, String> parent;
    private Map<String, String> emailToName;
    private Map<String, List<String>> groups;

    public List<List<String>> accountsMerge(List<List<String>> accounts) {
        parent = new HashMap<>();
        emailToName = new HashMap<>();
        groups = new HashMap<>();

        for (List<String> account : accounts) {
            String name = account.get(0);

            for (int i = 1; i < account.size(); i++) {
                String email = account.get(i);
                parent.putIfAbsent(email, email);
                emailToName.put(email, name);

                if (i > 1) {
                    union(account.get(1), email);
                }
            }
        }

        for (String email : parent.keySet()) {
            String root = find(email);
            groups.computeIfAbsent(root, k -> new ArrayList<>()).add(email);
        }

        List<List<String>> result = new ArrayList<>();
        for (String root : groups.keySet()) {
            List<String> emails = groups.get(root);
            Collections.sort(emails);

            List<String> account = new ArrayList<>();
            account.add(emailToName.get(root));
            account.addAll(emails);

            result.add(account);
        }

        return result;
    }

    private String find(String x) {
        if (parent.get(x).equals(x)) return x;
        String root = find(parent.get(x));
        parent.put(x, root);
        return root;
    }

    private void union(String x, String y) {
        String rootX = find(x);
        String rootY = find(y);

        if (rootX != rootY) {
            parent.put(rootY, rootX);
        }
    }
}

File: 0721-accounts-merge/0721-accounts-merge.kt
class Solution {
    private val parent = mutableMapOf<String, String>()
    private val emailToName = mutableMapOf<String, String>()
    private val groups = mutableMapOf<String, MutableList<String>>()

    fun accountsMerge(accounts: List<List<String>>): List<List<String>> {
        for (account in accounts) {
            val name = account[0]

            for (i in 1 until account.size) {
                val email = account[i]
                parent.putIfAbsent(email, email)
                emailToName[email] = name

                if (i > 1) union(account[1], email)
            }
        }

        for (email in parent.keys) {
            val root = find(email)
            groups.getOrPut(root) { mutableListOf() }.add(email)
        }

        return groups.map { (root, emails) ->
            emails.sort()
            mutableListOf(emailToName[root]!!).apply {
                addAll(emails)
            }
        }
    }

    private fun find(x: String): String {
        if (parent[x] == x) return x
        val root = find(parent[x]!!)
        parent[x] = root
        return root
    }

    private fun union(x: String, y: String) {
        val rootX = find(x)
        val rootY = find(y)

        if (rootX != rootY) {
            parent[rootY] = rootX
        }
    }
}